### PR TITLE
Add `ErrorGuaranteed` to `ast::ExprKind::Err`

### PR DESCRIPTION
**Pull request by ShE3py**
See #119967 for context
```
      \
       \
          _~^~^~_
      \) /  o o  \ (/
        '_   -   _'
        / '-----' \
```

r? fmease

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:327]: ❤️

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (fmease) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:606]: I wonder if we should agree on a single naming scheme here. Currently you use `valid`, `validity` and `result`. Maybe use `validity` everywhere (only in the places you had to modify)?. Not required tho

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (fmease) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:667]: :tada:

File: compiler/rustc_session/src/errors.rs
Comment (fmease) [inline compiler/rustc_session/src/errors.rs:385]:
```suggestion
            dcx.span_delayed_bug(span, "lexer error encountered but no error was emitted")
```

or something like that (more descriptive and to make the comment redundant).

File: compiler/rustc_parse/src/parser/expr.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/expr.rs:3435]: I haven't reviewed this function yet as the changes look non-trivial and I'm a bit exhausted now.
**Note to self**: Review this later.

File: compiler/rustc_parse/src/parser/stmt.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/stmt.rs:739]: Oh, this dance looks a bit unfortunate

**Review by fmease — COMMENTED** [review]
Very nice, thanks for working on this!

I have a couple of minor suggestions but there are still some parser changes I need to review. ~~Note that I probably won't approve this until CI works again~~ (oh, it runs again!). The tree is closed anyway.

~~You also need to to update `src/tools/rustfmt`~~ (ah, you just pushed them haha). I see that you've already updated Clippy, amazing!

File: compiler/rustc_builtin_macros/src/concat_bytes.rs
Comment (ShE3py) [inline compiler/rustc_builtin_macros/src/concat_bytes.rs:65]:
```suggestion
/// updates `emitted_err` accordingly.
```
(:

File: compiler/rustc_expand/src/mbe/macro_rules.rs
Comment (ShE3py) [reply compiler/rustc_expand/src/mbe/macro_rules.rs]: I was unhappy with all my tries but I found `emitted_err` which is kinda what the variable represents; also replaced the `mut` ones with `Option<ErrorGuaranteed>` instead of `Result<(), ErrorGuaranteed>` (which I use for function as they may want to return something someday).

**Comment by ShE3py** [comment]
Refactored some code, also took the liberty to remove the `ast::` invasion in the builtin macros I modified.

@rustbot review

File: src/tools/clippy/clippy_utils/src/ast_utils.rs
Comment (flip1995) [inline src/tools/clippy/clippy_utils/src/ast_utils.rs:146]: `(Dummy, Dummy)` is missing from this match statement.

Comment (ShE3py) [reply src/tools/clippy/clippy_utils/src/ast_utils.rs:146]: To mimic the old behavior `(Err(_), Dummy)` and `(Dummy, Err(_))` may also be needed, and I wonder if `(Dummy, _) | (_, Dummy) => true` makes sense.

But the only two occurences of `ExprKind::Dummy` are in
- `<Expr as DummyAstNode>::dummy()` which isn't actually called (I can replace its body with `panic!()` and both `./x test --stage 1 tests/ui/` and `./x check --stage 2` passes successfully)
  https://github.com/rust-lang/rust/blob/8c0b4f67c9b4bf477df38c16669fd576d46a2b3d/compiler/rustc_ast/src/mut_visit.rs#L1635-L1645
- `Parser::maybe_recover_trailing_expr()` for avoiding creating a `P<Expr>`
  https://github.com/rust-lang/rust/blob/8c0b4f67c9b4bf477df38c16669fd576d46a2b3d/compiler/rustc_parse/src/parser/pat.rs#L388-L394

So `ast_utils.rs`'s `eq_expr` cannot actually have an `ExprKind::Dummy`, so this case can't be tested in the testsuite. I was tempted to put a `panic!()` to hold the choice until later if someone uses `Dummy` and it reaches Clippy, but if it's not tested I'd rather have a wrong diagnostic than a panic in prod.

**Comment by fmease** [comment]
Just to give an update: I'm at 35 / 39 file viewed with pending review comments. Finishing my review later.

File: compiler/rustc_expand/src/base.rs
Comment (nnethercote) [inline compiler/rustc_expand/src/base.rs:556]: Would `Option<ErrorGuaranteed>` be better than `Result<(), ErrorGuaranteed>`?

**Comment by nnethercote** [comment]
Oh dear, I started working on this today, not realizing that someone else was working on it. I also did `TyKind::Err` (#121072) and `LitKind::Err` (no PR yet).

I didn't get all the way to fixing `ErrKind::Err`, but I think it's possible to do this without introducing `ErrKind::Dummy`, which would greatly reduce the size of the diff.

File: compiler/rustc_session/src/errors.rs
Comment (nnethercote) [inline compiler/rustc_session/src/errors.rs:390]: You can just `unwrap`.

**Review by fmease — COMMENTED** [review]
Just posting my pending review comments. I still need to look at some of these files.

**Comment by fmease** [comment]
> […] I think it's possible to do this without introducing ErrKind::Dummy, which would reduce the size of the diff.

@nnethercote How would you go about updating trait `DummyAstNode` and everything that uses it without `ExprKind::Dummy`?

File: compiler/rustc_expand/src/base.rs
Comment (ShE3py) [reply compiler/rustc_expand/src/base.rs:556]: I hesitated between `Option` and `Result` in several places, and I've just realized now that it's actually just `bool` → `Option<ErrorGuaranteed>` and `T` → `Result<T, ErrorGuaranteed>` conversions, only that certain functions had an implicit `()` returns.

I'm busy for the next few days so I'll probably do this (and the next suggestions) on Monday.

**Comment by nnethercote** [comment]
> @nnethercote How would you go about updating trait `DummyAstNode` and everything that uses it without `ExprKind::Dummy`?

Use `ExprKind::Tup(ThinVec::new())` wherever possible, and only use `ExprKind::Err` if absolutely necessary. `DummyAstNode` for `PatKind` and (soon) `TyKind` use a non-error value. And `DummyResult::raw_expr` is another example that uses `ExprKind::Tup(ThinVec::new())` where possible.

Because I didn't fully convert `ExprKind::Err` myself I'm not 100% sure, but it really seems like "use `Tup(ThinVec::new())` wherever possible and `Err(guar)` everywhere else" should be enough.

File: compiler/rustc_builtin_macros/src/concat.rs
Comment (nnethercote) [inline compiler/rustc_builtin_macros/src/concat.rs:38]: I would rename `emitted_err` as `guar` (because it's a guaranteed, not an error) and then just do `guar = Some(cx.dcx().emit_err(...))` in a single  statement.

File: compiler/rustc_builtin_macros/src/concat.rs
Comment (nnethercote) [inline compiler/rustc_builtin_macros/src/concat.rs:45]: This delayed bug won't be necessary after #121120.

File: compiler/rustc_expand/src/base.rs
Comment (nnethercote) [inline compiler/rustc_expand/src/base.rs:1282]: This delayed bug won't be necessary after https://github.com/rust-lang/rust/pull/121120.

File: compiler/rustc_builtin_macros/src/concat_bytes.rs
Comment (nnethercote) [inline compiler/rustc_builtin_macros/src/concat_bytes.rs:46]: This delayed bug won't be necessary after https://github.com/rust-lang/rust/pull/121120.

File: compiler/rustc_builtin_macros/src/concat.rs
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/concat.rs:38]: Likewise for other places where `emitted_err` is used as a name.

File: src/tools/clippy/clippy_utils/src/ast_utils.rs
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/ast_utils.rs:146]: If this can really not occur here, then I suggest to add a `debug_assertion` in this function. I'd like this to be an exhaustive match.

**Review by fmease — APPROVED** [review]
Thank you so much for your patience, I've finally managed to meticulously look through every single file.

Overall the changes look good. I have a couple of suggestions.
I agree with nnethercote that you should rename `emitted_err` to `guar`.
Just to clear up any potential confusion, we're gonna keep using `ExprKind::Dummy` (for now) as discussed (e.g., over `ExprKind::Tup(ThinVec::new())`).

This is good to go after a rebase and after some commit squashing. No need to squash it into a single commit, just fewer (you can probably git-fixup “address review comments” commits). And you need to get rid of that one merge commit you have (we have a no-merge commit policy).

**Comment by ShE3py** [comment]
Renamed `emitted_err` → `guar`, needed to rename some pats to do so:
```rs
ExprKind::Err(guarantee) => {
    guar = Some(*guarantee);
}
```
Diff: <https://github.com/ShE3py/rust/compare/bak-exprkind-err..exprkind-err>

I'll continue after sleeping :zzz:

File: compiler/rustc_expand/src/base.rs
Comment (ShE3py) [reply compiler/rustc_expand/src/base.rs:1075]: https://github.com/rust-lang/rust/blob/8c0b1fcd2914caaf1c3a1071028fb74b70c519e9/compiler/rustc_expand/src/expand.rs#L639-L648

The error is only emitted once (`MacroExpander::error_recursion_limit_reached`) and `AstFragmentKind::dummy` requires a `guar`.

It's possible to use `DiagCtxt::has_errors().unwrap()`, but we lose the guarantee that if the recursion limit was reduced, then an error was emitted.

File: compiler/rustc_expand/src/base.rs
Comment (ShE3py) [reply compiler/rustc_expand/src/base.rs]: Replaced with `span_bug`, this isn't currently reachable. I guess I used the delayed version 'cause my IDE suggested that and I didn't think too much about it, but `ExprKind::Dummy` only uses `span_bug` (to favor `ExprKind::Err`).

**Review by fmease — APPROVED** [review]
Thanks!

**Comment by fmease** [comment]
> FAILED: lib/liblldMachO.a
> C:\a\rust\rust\mingw64\bin\ranlib.exe: could not create temporary file whilst writing archive: no more archived files

@bors retry (spurious?)

**Pull request merged**
